store/tikv/2pc_exec.go

```diff
@@ -0,0 +1,394 @@
+	cancel()
+	return errors.Trace(err)
+}
+
+func (c *execController) handlePrimaryMutation(actionIsCommit bool) (int, error) {
+	var batch *batchMutations
+	for {
+		collector, err := c.getCollector(c.bo, c.committer.primaryKey, nil)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		batch, err = collector.Collect()
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		retry, err := c.action.handleSingleBatch(c.committer, c.bo, batch)
+		if err == nil && !retry {
+			break
+		}
+		if !retry {
+			return 0, errors.Trace(err)
+		}
+		if err != nil {
+			err = c.bo.Backoff(BoRegionMiss, err)
+			if err != nil {
+				return 0, errors.Trace(err)
+			}
+		}
+	}
+
+	knobs := &c.committer.testingKnobs
+	if actionIsCommit && knobs.bkAfterCommitPrimary != nil && knobs.acAfterCommitPrimary != nil {
+		knobs.acAfterCommitPrimary <- struct{}{}
+		<-knobs.bkAfterCommitPrimary
+	}
+
+	start, end := batch.mutations.keys[0], batch.mutations.keys[batch.mutations.len()-1]
+	c.filter = func(key []byte) bool {
+		return bytes.Compare(key, start) < 0 || bytes.Compare(key, end) > 0
+	}
+	return batch.mutations.len(), nil
+}
+
+func (c *execController) handleMutations() (err error) {
+	collector, err := c.getCollector(c.bo, nil, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var prev RegionVerID
+	for {
+		var batch *batchMutations
+		batch, err = collector.Collect()
+
+		if err != nil {
+			c.handleError(err)
+			break
+		}
+		if batch == nil {
+			break
+		}
+		if batch.region != prev {
+			prev = batch.region
+			c.numRegions++
+		}
+
+		if collector.Finished() {
+			bo, cancel := c.bo.Fork()
+			c.handleSingleBatch(&execJob{batch, bo, cancel})
+			break
+		}
+
+		if err = c.dispatchBatch(c.bo, batch); err != nil {
+			break
+		}
+	}
+	c.Wait()
+	if err == nil {
+		select {
+		case err = <-c.errCh:
+		default:
+		}
+	}
+	close(c.jobCh)
+
+	return errors.Trace(err)
+}
+
+func (c *execController) containsPrimaryKey() bool {
+	it := c.mutations.Iter(c.committer.primaryKey, nil)
+	m := it.Next()
+	return bytes.Equal(m.key, c.committer.primaryKey)
+}
+
+func (c *execController) getCollector(bo *Backoffer, start, end []byte) (*mutationBatchCollector, error) {
+	_, actionIsPrewrite := c.action.(actionPrewrite)
+	it := c.mutations.Iter(start, end)
+	if c.filter != nil {
+		it.WithFilter(c.filter)
+	}
+
+	return c.committer.newBatchCollector(bo, it, txnCommitBatchSize, actionIsPrewrite)
+}
+
+func (c *execController) dispatchBatch(bo *Backoffer, batch *batchMutations) error {
+	fbo, cancel := bo.Fork()
+	job := &execJob{batch, fbo, cancel}
+	c.Add(1)
+
+	select {
+	case c.jobCh <- job:
+		return nil
+	default:
+		c.tryAddWorker()
+	}
+
+	if c.cancel == nil {
+		c.jobCh <- job
+		return nil
+	}
+
+	// We can cancel dispatcher by return error.
+	select {
+	case c.jobCh <- job:
+		return nil
+	case err := <-c.errCh:
+		c.Done()
+		return err
+	}
+}
+
+func (c *execController) workerLoop() {
+	for job := range c.jobCh {
+		if !c.terminated() {
```

Comment:
I'm a bit worried about the worker's goroutine leak.
`c.terminated` is an edge trigger condition, if this line miss and the code runs to the for-range loop, it may block forever.